idx/memory/tag_query.go

```diff
@@ -2,504 +2,240 @@ package memory
-		startWith: query.StartWith,
-		from:      query.From,
-	}
+	return ctx
 }
 
-// getInitialByEqual generates the initial resultset by executing the given equal expression
-func (q *TagQueryContext) getInitialByEqual(expr kv, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
-
-KEYS:
-	for k := range q.index[expr.Key][expr.Value] {
-		select {
-		case <-stopCh:
-			break KEYS
-		case idCh <- k:
-		}
+func (q *TagQueryContext) prepareExpressions(idx TagIndex) {
+	type expressionCost struct {
+		cost          uint32
+		expressionIdx int
 	}
+	costs := make([]expressionCost, len(q.query.Expressions))
 
-	close(idCh)
-}
+	for i, expr := range q.query.Expressions {
+		costs[i].expressionIdx = i
 
-// getInitialByPrefix generates the initial resultset by executing the given prefix match expression
-func (q *TagQueryContext) getInitialByPrefix(expr kv, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
-
-VALUES:
-	for v, ids := range q.index[expr.Key] {
-		if !strings.HasPrefix(v, expr.Value) {
+		keyPtr, found := expr.GetKeyPtr()
+		if !found {
+			// the key of this expression has not been interned, this means it cannot exist in the index.
+			// so this expression will be cheap to evaluate
+			costs[i].cost = 0
 			continue
 		}
 
-		for id := range ids {
-			select {
-			case <-stopCh:
-				break VALUES
-			case idCh <- id:
+		if expr.ValueMatchesExactly() {
+			valuePtr, found := expr.GetValuePtr()
+			if !found {
+				// the key of this expression has not been interned, this means it cannot exist in the index.
+				// so this expression will be cheap to evaluate
+				costs[i].cost = 0
+				continue
+			}
+
+			costs[i].cost = uint32(len(idx[keyPtr][valuePtr])) * expr.GetCostMultiplier()
+		} else {
+			if expr.OperatesOnTag() {
+				costs[i].cost = uint32(len(idx)) * expr.GetCostMultiplier()
+			} else {
+				costs[i].cost = uint32(len(idx[keyPtr])) * expr.GetCostMultiplier()
 			}
 		}
 	}
 
-	close(idCh)
-}
+	sort.Slice(costs, func(i, j int) bool { return costs[i].cost < costs[j].cost })
 
-// getInitialByMatch generates the initial resultset by executing the given match expression
-func (q *TagQueryContext) getInitialByMatch(expr kvRe, idCh chan schema.MKey, stopCh chan struct{}) {
-	defer q.wg.Done()
+	q.filters = make([]tagquery.MetricDefinitionFilter, len(q.query.Expressions)-1)
+	q.defaultDecisions = make([]tagquery.FilterDecision, len(q.query.Expressions)-1)
 
-	// shortcut if Regex == nil.
-	// this will simply match any value, like ^.+. since we know that every value
-	// in the index must not be empty, we can skip the matching.
-	if expr.Regex == nil {
-	VALUES1:
-		for _, ids := range q.index[expr.Key] {
-			for id := range ids {
-				select {
-				case <-stopCh:
-					break VALUES1
-				case idCh <- id:
-				}
-			}
+	i := 0
+	for _, cost := range costs {
+		if q.startWith < 0 && q.query.Expressions[cost.expressionIdx].RequiresNonEmptyValue() {
+			q.startWith = cost.expressionIdx
```

Comment:
what's going on here? are we just setting the startWith to the first expression that requires a non empty value?
what if we don't find one? (and startWith remains < 0) ?

Reply:
the expressions have at that point already been sorted by cost. so yes, we just take the first one that requires a non-empty value. 
every set of expressions must include at least one expression requiring a non-empty value according to https://graphite.readthedocs.io/en/latest/tags.html#querying
`tagquery.NewQuery` enforces this rule, here we trust that the tag queries which get passed into the index have already been validated. If somebody instantiates a tag query without using `tagquery.NewQuery` then it's their responsibility to ensure it is valid. I'll add a comment mentioning this.